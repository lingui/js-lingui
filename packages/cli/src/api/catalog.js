// @flow
import fs from "fs"
import path from "path"
import mkdirp from "mkdirp"
import glob from "glob"
import R from "ramda"

import type { LinguiConfig, CatalogApi } from "./types"
import * as locales from "./locales"
import { formats } from "."

export default (config: LinguiConfig): CatalogApi => {
  const format = formats[config.format]
  if (!format) {
    throw new Error(`Unknown format ${config.format}.`)
  }

  const sourceFilename = path.join("{locale}", format.filename)
  const compiledFilename = path.join("{locale}", "messages.js")

  return {
    formatFilename(pattern, locale) {
      return pattern.replace("{locale}", locale)
    },

    write(locale, messages) {
      const filename = path.join(
        config.localeDir,
        this.formatFilename(sourceFilename, locale)
      )

      const created = !fs.existsSync(filename)
      format.write(filename, messages, { language: locale })
      return [created, filename]
    },

    read(locale) {
      const filename = path.join(
        config.localeDir,
        this.formatFilename(sourceFilename, locale)
      )

      if (!fs.existsSync(filename)) return null
      return format.read(filename)
    },

    readAll() {
      return R.mergeAll(
        this.getLocales().map(locale => ({
          [locale]: this.read(locale)
        }))
      )
    },

    merge(prevCatalogs, nextCatalog, options = {}) {
      const nextKeys = R.keys(nextCatalog)

      return R.mapObjIndexed((prevCatalog, locale) => {
        const prevKeys = R.keys(prevCatalog)

        const newKeys = R.difference(nextKeys, prevKeys)
        const mergeKeys = R.intersection(nextKeys, prevKeys)
        const obsoleteKeys = R.difference(prevKeys, nextKeys)

        // Initialize new catalog with new keys
        const newMessages = R.mapObjIndexed(
          (message, key) => ({
            translation:
              config.sourceLocale === locale ? message.defaults || key : "",
            ...message
          }),
          R.pick(newKeys, nextCatalog)
        )

        // Merge translations from previous catalog
        const mergedMessages = mergeKeys.map(key => {
          const updateFromDefaults =
            config.sourceLocale === locale &&
            (prevCatalog[key].translation === prevCatalog[key].defaults ||
              options.overwrite)

          const translation = updateFromDefaults
            ? nextCatalog[key].defaults
            : prevCatalog[key].translation

          return {
            [key]: {
              translation,
              ...R.omit(["obsolete, translation"], nextCatalog[key])
            }
          }
        })

        // Mark all remaining translations as obsolete
        const obsoleteMessages = obsoleteKeys.map(key => ({
          [key]: {
            ...prevCatalog[key],
            obsolete: true
          }
        }))

        return R.mergeAll([newMessages, ...mergedMessages, ...obsoleteMessages])
      }, prevCatalogs)
    },

    getTranslation(catalogs, locale, key, { fallbackLocale, sourceLocale }) {
      const getTranslation = locale => catalogs[locale][key].translation

      return (
        // Get translation in target locale
        getTranslation(locale) ||
        // Get translation in fallbackLocale (if any)
        (fallbackLocale && getTranslation(fallbackLocale)) ||
        // Get message default
        catalogs[locale][key].defaults ||
        // If sourceLocale is either target locale of fallback one, use key
        (sourceLocale && sourceLocale === locale && key) ||
        (sourceLocale &&
          fallbackLocale &&
          sourceLocale === fallbackLocale &&
          key) ||
        // Otherwise no translation is available
        undefined
      )
    },

    writeCompiled(locale, content) {
      const filename = path.join(
        config.localeDir,
        this.formatFilename(compiledFilename, locale)
      )

      fs.writeFileSync(filename, content)
      return filename
    },

    getLocale(filename) {
      const [messages, locale] = filename.split(path.sep).reverse()

      if (messages !== format.filename || !locales.isValid(locale)) return null
      return locale
    },

    getLocales() {
      const pattern = path.join(config.localeDir, "*")

      return glob
        .sync(pattern)
        .map(filename => {
          // Don't use path.sep here, because glob.sync normalizes path separators
          const [locale] = filename.split("/").reverse()
          return (
            (locales.isValid(locale) || locale === config.pseudoLocale) &&
            locale
          )
        })
        .filter(Boolean)
    },

    addLocale(locale) {
      if (!locales.isValid(locale) && locale !== config.pseudoLocale) {
        return [false, null]
      }

      const filename = path.join(
        config.localeDir,
        this.formatFilename(sourceFilename, locale)
      )

      if (!fs.existsSync(filename)) {
        const dirname = path.dirname(filename)

        mkdirp.sync(dirname)
        this.write(locale, {})

        return [true, filename]
      }

      return [false, filename]
    }
  }
}
