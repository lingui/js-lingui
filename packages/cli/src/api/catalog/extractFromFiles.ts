import type {
  ExtractedMessage,
  ExtractorType,
  LinguiConfigNormalized,
} from "@lingui/conf"
import extract from "../extractors"
import path from "path"
import chalk from "chalk"
import { prettyOrigin } from "../utils"
import { MessageOrigin, ExtractedCatalogType } from "../types"

export async function extractFromFiles(
  paths: string[],
  config: LinguiConfigNormalized
) {
  const messages: ExtractedCatalogType = {}

  let catalogSuccess = true
  for (let filename of paths) {
    const fileSuccess = await extract(
      filename,
      (next: ExtractedMessage) => {
        if (!messages[next.id]) {
          messages[next.id] = {
            message: next.message,
            context: next.context,
            comments: [],
            origin: [],
          }
        }

        const prev = messages[next.id]

        // there might be a case when filename was not mapped from sourcemaps
        const filename = next.origin[0]
          ? path.relative(config.rootDir, next.origin[0]).replace(/\\/g, "/")
          : ""

        const origin: MessageOrigin = [filename, next.origin[1]]

        if (prev.message && next.message && prev.message !== next.message) {
          throw new Error(
            `Encountered different default translations for message ${chalk.yellow(
              next.id
            )}` +
              `\n${chalk.yellow(prettyOrigin(prev.origin))} ${prev.message}` +
              `\n${chalk.yellow(prettyOrigin([origin]))} ${next.message}`
          )
        }

        messages[next.id] = {
          ...prev,
          comments: next.comment
            ? [...prev.comments, next.comment]
            : prev.comments,
          origin: [...prev.origin, [filename, next.origin[1]]],
        }
      },
      config,
      {
        extractors: config.extractors as ExtractorType[],
      }
    )
    catalogSuccess &&= fileSuccess
  }

  if (!catalogSuccess) return undefined

  return messages
}
