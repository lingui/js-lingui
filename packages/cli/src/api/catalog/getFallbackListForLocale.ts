import { FallbackLocales } from "@lingui/conf"

export function getFallbackListForLocale(
  fallbackLocales: FallbackLocales,
  locale: string
): string[] {
  const fL: string[] = []

  if (fallbackLocales?.[locale]) {
    const mapping = fallbackLocales?.[locale]
    Array.isArray(mapping) ? fL.push(...mapping) : fL.push(mapping)
  }

  if (fallbackLocales?.default && locale !== fallbackLocales?.default) {
    fL.push(fallbackLocales?.default)
  }

  return fL
}
