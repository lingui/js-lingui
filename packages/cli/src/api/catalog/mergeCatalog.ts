import * as R from "ramda"
import type { MergeOptions } from "../catalog"
import { CatalogType, ExtractedCatalogType, MessageType } from "../types"

export function mergeCatalog(
  prevCatalog: CatalogType,
  nextCatalog: ExtractedCatalogType,
  forSourceLocale: boolean,
  options: MergeOptions
): CatalogType {
  const nextKeys = Object.keys(nextCatalog)

  const prevKeys = R.keys(prevCatalog).map(String)

  const newKeys = R.difference(nextKeys, prevKeys)
  const mergeKeys = R.intersection(nextKeys, prevKeys)
  const obsoleteKeys = R.difference(prevKeys, nextKeys)

  // Initialize new catalog with new keys
  const newMessages = R.mapObjIndexed(
    (message: MessageType, key) => ({
      translation: forSourceLocale ? message.message || key : "",
      ...message,
    }),
    R.pick(newKeys, nextCatalog)
  )

  // Merge translations from previous catalog
  const mergedMessages = mergeKeys.map((key) => {
    const updateFromDefaults =
      forSourceLocale &&
      (prevCatalog[key].translation === prevCatalog[key].message ||
        options.overwrite)

    const translation = updateFromDefaults
      ? nextCatalog[key].message || key
      : prevCatalog[key].translation

    return {
      [key]: {
        translation,
        ...R.omit(["obsolete, translation"], nextCatalog[key]),
      },
    }
  })

  // Mark all remaining translations as obsolete
  // Only if *options.files* is not provided
  const obsoleteMessages = obsoleteKeys.map((key) => ({
    [key]: {
      ...prevCatalog[key],
      obsolete: !options.files,
    },
  }))

  return R.mergeAll([newMessages, ...mergedMessages, ...obsoleteMessages])
}
