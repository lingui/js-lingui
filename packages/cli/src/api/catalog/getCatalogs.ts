import { LinguiConfigNormalized } from "@lingui/conf"
import glob from "glob"
import path from "path"
import { Catalog } from "../catalog"
import { normalizeRelativePath, PATHSEP, replacePlaceholders } from "../utils"
import micromatch from "micromatch"
import { getFormat } from "../formats"
import { getExperimentalCatalogs } from "../../extract-experimental/getExperimentalCatalogs"

const NAME_PH = "{name}"
const LOCALE_PH = "{locale}"

/**
 * Parse `config.catalogs` and return a list of configured Catalog instances.
 */
export async function getCatalogs(
  config: LinguiConfigNormalized
): Promise<Catalog[]> {
  const catalogsConfig = config.catalogs
  const catalogs: Catalog[] = []

  const format = await getFormat(
    config.format,
    config.formatOptions,
    config.sourceLocale
  )

  catalogsConfig.forEach((catalog) => {
    validateCatalogPath(catalog.path, format.getCatalogExtension())

    const include = ensureArray(catalog.include).map(normalizeRelativePath)
    const exclude = ensureArray(catalog.exclude).map(normalizeRelativePath)

    // catalog.path without {name} pattern -> always refers to a single catalog
    if (!catalog.path.includes(NAME_PH)) {
      // Validate that sourcePaths doesn't use {name} pattern either
      const invalidSource = include.find((path) => path.includes(NAME_PH))
      if (invalidSource !== undefined) {
        throw new Error(
          `Catalog with path "${catalog.path}" doesn't have a {name} pattern` +
            ` in it, but one of source directories uses it: "${invalidSource}".` +
            ` Either add {name} pattern to "${catalog.path}" or remove it` +
            ` from all source directories.`
        )
      }

      catalogs.push(
        new Catalog(
          {
            name: getCatalogName(catalog.path),
            path: normalizeRelativePath(catalog.path),
            include,
            exclude,
            format,
          },
          config
        )
      )
      return
    }

    const patterns = include.map((path) =>
      replacePlaceholders(path, { name: "*" })
    )
    const candidates = glob.sync(
      patterns.length > 1 ? `{${patterns.join(",")}}` : patterns[0],
      {
        ignore: exclude,
        mark: true,
      }
    )

    candidates.forEach((catalogDir) => {
      const name = path.basename(catalogDir)
      catalogs.push(
        new Catalog(
          {
            name,
            path: normalizeRelativePath(
              replacePlaceholders(catalog.path, { name })
            ),
            include: include.map((path) => replacePlaceholders(path, { name })),
            exclude: exclude.map((path) => replacePlaceholders(path, { name })),
            format,
          },
          config
        )
      )
    })
  })

  if (config.experimental?.extractor?.entries.length) {
    catalogs.push(...(await getExperimentalCatalogs(config)))
  }

  return catalogs
}

/**
 * Ensure that value is always array. If not, turn it into an array of one element.
 */
const ensureArray = <T>(value: Array<T> | T | null | undefined): Array<T> => {
  if (value == null) return []

  return Array.isArray(value) ? value : [value]
}

/**
 * Create catalog for merged messages.
 */
export async function getCatalogForMerge(config: LinguiConfigNormalized) {
  const format = await getFormat(
    config.format,
    config.formatOptions,
    config.sourceLocale
  )
  validateCatalogPath(config.catalogsMergePath, format.getCatalogExtension())

  return new Catalog(
    {
      name: getCatalogName(config.catalogsMergePath),
      path: normalizeRelativePath(config.catalogsMergePath),
      include: [],
      exclude: [],
      format,
    },
    config
  )
}

export function getCatalogForFile(file: string, catalogs: Catalog[]) {
  for (const catalog of catalogs) {
    const catalogFile = `${catalog.path}${catalog.format.getCatalogExtension()}`
    const catalogGlob = replacePlaceholders(catalogFile, { locale: "*" })
    const match = micromatch.capture(
      normalizeRelativePath(path.relative(catalog.config.rootDir, catalogGlob)),
      normalizeRelativePath(file)
    )
    if (match) {
      return {
        locale: match[0],
        catalog,
      }
    }
  }

  return null
}

/**
 *  Validate that `catalogPath` doesn't end with trailing slash
 */
function validateCatalogPath(path: string, extension: string) {
  if (!path.endsWith(PATHSEP)) {
    return
  }

  const correctPath = path.slice(0, -1)
  const examplePath =
    replacePlaceholders(correctPath, {
      locale: "en",
    }) + extension
  throw new Error(
    // prettier-ignore
    `Remove trailing slash from "${path}". Catalog path isn't a directory,` +
    ` but translation file without extension. For example, catalog path "${correctPath}"` +
    ` results in translation file "${examplePath}".`
  )
}

function getCatalogName(filePath: string) {
  // catalog name is the last directory of catalogPath.
  // If the last part is {locale}, then catalog doesn't have an explicit name
  const _name = path.basename(normalizeRelativePath(filePath))
  return _name !== LOCALE_PH ? _name : null
}
