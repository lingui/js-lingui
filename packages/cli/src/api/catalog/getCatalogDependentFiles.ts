import { Catalog } from "../catalog"
import { getFallbackListForLocale } from "./getFallbackListForLocale"
import fs from "node:fs/promises"

import path from "node:path"
import * as process from "process"

const fileExists = async (path: string) =>
  !!(await fs.stat(path).catch(() => false))

/**
 * Return all files catalog implicitly depends on.
 */
export async function getCatalogDependentFiles(
  catalog: Catalog,
  locale: string
): Promise<string[]> {
  const files = new Set([catalog.templateFile])
  getFallbackListForLocale(catalog.config.fallbackLocales, locale).forEach(
    (locale) => {
      files.add(catalog.getFilename(locale))
    }
  )

  if (catalog.config.sourceLocale && locale !== catalog.config.sourceLocale) {
    files.add(catalog.getFilename(catalog.config.sourceLocale))
  }

  const out: string[] = []

  for (let file of files) {
    file = path.isAbsolute(file) ? file : path.join(process.cwd(), file)

    if (await fileExists(file)) {
      out.push(file)
    }
  }

  return out
}
