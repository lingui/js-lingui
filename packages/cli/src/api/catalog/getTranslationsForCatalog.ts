import { Catalog } from "../catalog"
import { FallbackLocales } from "@lingui/conf"
import type { AllCatalogsType, CatalogType, MessageType } from "../types"
import { getFallbackListForLocale } from "./getFallbackListForLocale"

export type TranslationMissingEvent = {
  source: string
  id: string
}
export type GetTranslationsOptions = {
  sourceLocale: string
  fallbackLocales: FallbackLocales
  onMissing?: (message: TranslationMissingEvent) => void
}

export async function getTranslationsForCatalog(
  catalog: Catalog,
  locale: string,
  options: GetTranslationsOptions
) {
  const catalogs = await catalog.readAll()
  const template = (await catalog.readTemplate()) || {}

  const sourceLocaleCatalog = catalogs[options.sourceLocale] || {}

  const input = { ...template, ...sourceLocaleCatalog, ...catalogs[locale] }

  return Object.keys(input).reduce<{ [id: string]: string }>((acc, key) => {
    acc[key] = getTranslation(catalogs, input[key], locale, key, options)
    return acc
  }, {})
}

function sourceLocaleFallback(catalog: CatalogType, key: string) {
  if (!catalog?.[key]) {
    return null
  }

  return catalog[key].translation || catalog[key].message
}

function getTranslation(
  catalogs: AllCatalogsType,
  msg: MessageType,
  locale: string,
  key: string,
  options: GetTranslationsOptions
) {
  const { fallbackLocales, sourceLocale, onMissing } = options

  const getTranslation = (_locale: string) => {
    const localeCatalog = catalogs[_locale]
    return localeCatalog?.[key]?.translation
  }

  const getMultipleFallbacks = (_locale: string) => {
    const fL = getFallbackListForLocale(fallbackLocales, _locale)

    if (!fL.length) return null

    for (const fallbackLocale of fL) {
      if (catalogs[fallbackLocale] && getTranslation(fallbackLocale)) {
        return getTranslation(fallbackLocale)
      }
    }
  }

  // target locale -> fallback locales -> fallback locales default ->
  // ** (following fallbacks would emit `missing` warning) **
  // -> source locale translation -> source locale message
  // -> template message
  // ** last resort **
  // -> id
  let translation =
    // Get translation in target locale
    getTranslation(locale) ||
    // We search in fallbackLocales as dependent of each locale
    getMultipleFallbacks(locale) ||
    (sourceLocale &&
      sourceLocale === locale &&
      sourceLocaleFallback(catalogs[sourceLocale], key))

  if (!translation) {
    onMissing &&
      onMissing({
        id: key,
        source:
          msg.message || sourceLocaleFallback(catalogs[sourceLocale], key),
      })
  }

  return (
    translation ||
    (sourceLocale && sourceLocaleFallback(catalogs[sourceLocale], key)) ||
    // take from template
    msg.message ||
    key
  )
}
