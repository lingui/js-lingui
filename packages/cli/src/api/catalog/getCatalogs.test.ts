import mockFs from "mock-fs"
import {
  getCatalogForFile,
  getCatalogForMerge,
  getCatalogs,
} from "./getCatalogs"
import { Catalog } from "../catalog"
import { LinguiConfig, makeConfig } from "@lingui/conf"
import path from "path"
import { getFormat } from "@lingui/cli/api"
import { FormatterWrapper } from "../formats"

function mockConfig(config: Partial<LinguiConfig> = {}) {
  return makeConfig(
    {
      rootDir: path.join(__dirname, "fixtures"),
      locales: ["en", "pl"],
      ...config,
    },
    { skipValidation: true }
  )
}

describe("getCatalogs", () => {
  let format: FormatterWrapper

  beforeAll(async () => {
    format = await getFormat("po", {}, "en")
  })

  afterEach(() => {
    mockFs.restore()
  })

  it("should get single catalog if catalogPath doesn't include {name} pattern", async () => {
    const config = mockConfig({
      catalogs: [
        {
          path: "./src/locales/{locale}",
          include: ["./src/"],
        },
      ],
    })
    expect(cleanCatalog((await getCatalogs(config))[0])).toEqual(
      cleanCatalog(
        new Catalog(
          {
            name: null,
            path: "src/locales/{locale}",
            include: ["src/"],
            exclude: [],
            format,
          },
          config
        )
      )
    )
  })

  it("should have catalog name and ignore patterns", async () => {
    const config = mockConfig({
      catalogs: [
        {
          path: "src/locales/{locale}/all",
          include: ["src/", "/absolute/path/"],
          exclude: ["node_modules/"],
        },
      ],
    })
    expect(cleanCatalog((await getCatalogs(config))[0])).toEqual(
      cleanCatalog(
        new Catalog(
          {
            name: "all",
            path: "src/locales/{locale}/all",
            include: ["src/", "/absolute/path/"],
            exclude: ["node_modules/"],
            format,
          },
          config
        )
      )
    )
  })

  it("should expand {name} for matching directories", async () => {
    mockFs({
      componentA: {
        "index.js": mockFs.file(),
      },
      componentB: {
        "index.js": mockFs.file(),
      },
    })

    const config = mockConfig({
      catalogs: [
        {
          path: "{name}/locales/{locale}",
          include: ["./{name}/"],
        },
      ],
    })

    const catalogs = await getCatalogs(config)

    expect([cleanCatalog(catalogs[0]), cleanCatalog(catalogs[1])]).toEqual([
      cleanCatalog(
        new Catalog(
          {
            name: "componentA",
            path: "componentA/locales/{locale}",
            include: ["componentA/"],
            exclude: [],
            format,
          },
          config
        )
      ),
      cleanCatalog(
        new Catalog(
          {
            name: "componentB",
            path: "componentB/locales/{locale}",
            include: ["componentB/"],
            exclude: [],
            format,
          },
          config
        )
      ),
    ])
  })

  it("should expand {name} multiple times in path", async () => {
    mockFs({
      componentA: {
        "index.js": mockFs.file(),
      },
    })

    const config = mockConfig({
      catalogs: [
        {
          path: "{name}/locales/{locale}/{name}_messages_{locale}",
          include: ["./{name}/"],
        },
      ],
    })
    expect(cleanCatalog((await getCatalogs(config))[0])).toEqual(
      cleanCatalog(
        new Catalog(
          {
            name: "componentA",
            path: "componentA/locales/{locale}/componentA_messages_{locale}",
            include: ["componentA/"],
            exclude: [],
            format,
          },
          config
        )
      )
    )
  })

  it("shouldn't expand {name} for ignored directories", async () => {
    mockFs({
      componentA: {
        "index.js": mockFs.file(),
      },
      componentB: {
        "index.js": mockFs.file(),
      },
    })

    const config = mockConfig({
      catalogs: [
        {
          path: "./{name}/locales/{locale}",
          include: ["./{name}/"],
          exclude: ["componentB/"],
        },
      ],
    })
    expect(cleanCatalog((await getCatalogs(config))[0])).toEqual(
      cleanCatalog(
        new Catalog(
          {
            name: "componentA",
            path: "componentA/locales/{locale}",
            include: ["componentA/"],
            exclude: ["componentB/"],
            format,
          },
          config
        )
      )
    )
  })

  it("should warn if catalogPath is a directory", async () => {
    await expect(
      getCatalogs(
        mockConfig({
          catalogs: [
            {
              path: "./locales/{locale}/",
              include: ["."],
            },
          ],
        })
      )
    ).rejects.toMatchSnapshot()

    // Use values from config in error message
    await expect(
      getCatalogs(
        mockConfig({
          locales: ["cs"],
          catalogs: [
            {
              path: "./locales/{locale}/",
              include: ["."],
            },
          ],
        })
      )
    ).rejects.toMatchSnapshot()
  })

  it("should warn about missing {name} pattern in catalog path", async () => {
    await expect(
      getCatalogs(
        mockConfig({
          catalogs: [
            {
              path: "./locales/{locale}",
              include: ["./{name}/"],
            },
          ],
        })
      )
    ).rejects.toMatchSnapshot()
  })
})

// remove non-serializable properties, which are not subject of a test
function cleanCatalog(catalog: Catalog) {
  delete catalog.config
  delete catalog.format
  return catalog
}
describe("getCatalogForMerge", () => {
  let format: FormatterWrapper

  beforeAll(async () => {
    format = await getFormat("po", {}, "en")
  })

  afterEach(() => {
    mockFs.restore()
  })

  it("should return catalog for merged messages", async () => {
    const config = mockConfig({
      catalogsMergePath: "locales/{locale}",
    })
    expect(cleanCatalog(await getCatalogForMerge(config))).toEqual(
      cleanCatalog(
        new Catalog(
          {
            name: null,
            path: "locales/{locale}",
            include: [],
            exclude: [],
            format,
          },
          config
        )
      )
    )
  })

  it("should return catalog with custom name for merged messages", async () => {
    const config = mockConfig({
      catalogsMergePath: "locales/{locale}/my/dir",
    })
    expect(cleanCatalog(await getCatalogForMerge(config))).toStrictEqual(
      cleanCatalog(
        new Catalog(
          {
            name: "dir",
            path: "locales/{locale}/my/dir",
            include: [],
            exclude: [],
            format,
          },
          config
        )
      )
    )
  })

  it("should throw error if catalogsMergePath ends with slash", async () => {
    const config = mockConfig({
      catalogsMergePath: "locales/{locale}/bad/path/",
    })
    expect.assertions(1)
    try {
      await getCatalogForMerge(config)
    } catch (e) {
      expect((e as Error).message).toBe(
        'Remove trailing slash from "locales/{locale}/bad/path/". Catalog path isn\'t a directory, but translation file without extension. For example, catalog path "locales/{locale}/bad/path" results in translation file "locales/en/bad/path.po".'
      )
    }
  })

  it("should throw error if {locale} is omitted from catalogsMergePath", async () => {
    const config = mockConfig({
      catalogsMergePath: "locales/bad/path",
    })
    expect.assertions(1)
    try {
      await getCatalogForMerge(config)
    } catch (e) {
      expect((e as Error).message).toBe(
        "Invalid catalog path: {locale} variable is missing"
      )
    }
  })
})
describe("getCatalogForFile", () => {
  let format: FormatterWrapper

  beforeAll(async () => {
    format = await getFormat("po", {}, "en")
  })

  it("should return null if catalog cannot be found", () => {
    const catalogs = [
      new Catalog(
        {
          name: null,
          path: "./src/locales/{locale}",
          include: ["./src/"],
          format,
        },
        mockConfig()
      ),
    ]

    expect(getCatalogForFile("./xyz/en.po", catalogs)).toBeNull()
  })

  it("should return matching catalog and locale if {locale} is present multiple times in path", () => {
    const catalog = new Catalog(
      {
        name: null,
        path: "./src/locales/{locale}/messages_{locale}",
        include: ["./src/"],
        format,
      },
      mockConfig({ format: "po", rootDir: "." })
    )
    const catalogs = [catalog]

    expect(
      getCatalogForFile("./src/locales/en/messages_en.po", catalogs)
    ).toEqual({
      locale: "en",
      catalog,
    })
  })

  it("should return matching catalog and locale", () => {
    const catalog = new Catalog(
      {
        name: null,
        path: "./src/locales/{locale}",
        include: ["./src/"],
        format,
      },
      mockConfig({ format: "po", rootDir: "." })
    )
    const catalogs = [catalog]

    expect(getCatalogForFile("./src/locales/en.po", catalogs)).toEqual({
      locale: "en",
      catalog,
    })
  })

  it("should work with Windows path delimiters", () => {
    const catalog = new Catalog(
      {
        name: null,
        path: ".\\src\\locales\\{locale}",
        include: ["./src/"],
        format,
      },
      mockConfig({ format: "po", rootDir: "." })
    )
    const catalogs = [catalog]

    expect(getCatalogForFile("src\\locales\\en.po", catalogs)).toEqual({
      locale: "en",
      catalog,
    })
  })
})
