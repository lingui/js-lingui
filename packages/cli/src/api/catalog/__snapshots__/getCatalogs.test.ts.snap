// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getCatalogs should warn about missing {name} pattern in catalog path 1`] = `[Error: Catalog with path "./locales/{locale}" doesn't have a {name} pattern in it, but one of source directories uses it: "{name}". Either add {name} pattern to "./locales/{locale}" or remove it from all source directories.]`;

exports[`getCatalogs should warn if catalogPath is a directory 1`] = `[Error: Remove trailing slash from "./locales/{locale}/". Catalog path isn't a directory, but translation file without extension. For example, catalog path "./locales/{locale}" results in translation file "./locales/en.po".]`;

exports[`getCatalogs should warn if catalogPath is a directory 2`] = `[Error: Remove trailing slash from "./locales/{locale}/". Catalog path isn't a directory, but translation file without extension. For example, catalog path "./locales/{locale}" results in translation file "./locales/en.po".]`;
